### DISCUSSION TITLE
Mobius Addition Batch for Hyperboloid model?

### DISCUSSION THREAD

**Discussion by AdarshMJ** [geoopt/geoopt#175] [Q&A]
Hi! 
So I have been trying to convert Hyperbolic Image Embeddings (https://arxiv.org/abs/1904.02239 ) from the Poincare ball model to the Lorentz model. The authors in the paper have used operations like -

```python
def _mobius_addition_batch(x, y, c):
    xy = _tensor_dot(x, y)  # B x C
    x2 = x.pow(2).sum(-1, keepdim=True)  # B x 1
    y2 = y.pow(2).sum(-1, keepdim=True)  # C x 1
    num = 1 + 2 * c * xy + c * y2.permute(1, 0)  # B x C
    num = num.unsqueeze(2) * x.unsqueeze(1)
    num = num + (1 - c * x2).unsqueeze(2) * y  # B x C x D
    denom_part1 = 1 + 2 * c * xy  # B x C
    denom_part2 = c ** 2 * x2 * y2.permute(1, 0)
    denom = denom_part1 + denom_part2
    res = num / (denom.unsqueeze(2) + 1e-5)
    return res
```

```python
def _hyperbolic_softmax(X, A, P, c):
    lambda_pkc = 2 / (1 - c * P.pow(2).sum(dim=1))
    k = lambda_pkc * torch.norm(A, dim=1) / torch.sqrt(c)
    mob_add = _mobius_addition_batch(-P, X, c)
    num = 2 * torch.sqrt(c) * torch.sum(mob_add * A.unsqueeze(1), dim=-1)
    denom = torch.norm(A, dim=1, keepdim=True) * (1 - c * mob_add.pow(2).sum(dim=2))
    logit = k.unsqueeze(1) * arsinh(num / denom)
    return logit.permute(1, 0)
```
I am not entirely sure what are the hyperboloid model equivalents of these are. My strategy was to use the same operations and then map the points to the hyperboloid model using the formula -

1+||w||^2/2*||w|*(|1-||w||). (from https://arxiv.org/pdf/1903.07307.pdf section 2.2)

But doesn't seem to be working. Any insights are welcome please!

**Comment by ferrine** [comment]
`mobius_addition_batch` is https://github.com/geoopt/geoopt/blob/master/geoopt/manifolds/stereographic/manifold.py#L223
softmax needs some research. How do authors motivate their softmax? There are multiple ways to think about it

**Reply by ferrine** [reply]
I think this softmax might be `https://github.com/geoopt/geoopt/blob/master/geoopt/manifolds/stereographic/manifold.py#L335` as in the first paper

**Comment by AdarshMJ** [comment]
Hi! thanks for the reply. I am currently doing this -

`
def _hyperbolic_softmax(X, A, P, c):
    lambda_pkc = 2 / (1 - c * P.pow(2).sum(dim=1))
    k = lambda_pkc * torch.norm(A, dim=1) / torch.sqrt(c)
    mob_add = _mobius_addition_batch(-P, X, c)
    num = 2 * torch.sqrt(c) * torch.sum(mob_add * A.unsqueeze(1), dim=-1)
    denom = torch.norm(A, dim=1, keepdim=True) * (1 - c * mob_add.pow(2).sum(dim=2))
    logit = k.unsqueeze(1) * arsinh(num / denom)
    logit1 = p2k(logit,c)
    logit2 = k2h(logit1,c)
    return logit2.permute(1, 0)
`

Using this conversion -

<img width="815" alt="Screenshot 2021-06-21 at 12 00 14" src="https://user-images.githubusercontent.com/24205200/122744415-3dd9c800-d288-11eb-96d1-e4a9dbeee041.png">

where -

`def p2k(x, c):
    denom = 1 + c * x.pow(2).sum(-1, keepdim=True)
    return 2 * x / denom`


`def k2p(x, c):
    denom = 1 + torch.sqrt(1 - c * x.pow(2).sum(-1, keepdim=True))
    return x / denom`

Does this make sense?

**Reply by AdarshMJ** [reply]
Calculate the Poincare_Mean and also Poincare distance, map the answer to Klein model and then map to hyperboloid. The training seems to be working, but I am not entirely sure if this is the right way to go about. I also seem to be getting slightly better results compared to the original implementation, at least for few epochs like 5 instead of running for 200 because I only have access to Google Colab :(

**Reply by ferrine** [reply]
There is some WIP here: https://github.com/geoopt/geoopt/pull/143

**Reply by ferrine** [reply]
You can find conversion functions as well

**Reply by AdarshMJ** [reply]
HI thanks for the reply. I will check these out! 
Can you please verify if my implementation of converting to to and fro is valid? Like does this idea make sense?
